### ISSUE TITLE
Converting macros to enums

### ISSUE THREAD

**Issue by IIFE** [mono/CppSharp#1377]
Hi,

I've been trying to use `GenerateEnumFromMacros` to generate enums from macros, however I've run into an issue when my macros are defined as follows:

```c++
#define D1 1
#define D2 2
#define D3 D1 | D2
```

Basically `D3` is an XOR of `D1` and `D2`, however CppSharp tries to parse the string `D1 | D2` as integer, which blows up.

Is it not possible to evaluate XOR expression for macros? Does the underlying `clang` parser not provide any extra details about macros?

Thanks

**Comment by ddobrev** [comment]
Clang provides everything but we don't read and handle everything. Macros in particular are too diverse and have to therefore be handled per case. So I'm afraid that if you want support for such defines, you'll have to write it yourself.

**Comment by IIFE** [comment]
@ddobrev So does clang provide the evaluated value for an XOR expression in a macro definition? If it does, wouldn't that be useful to expose in the C# MacroDefinition object? At least for numeric macros.

For example the above example, does clang provide a property that says the value for the `D3` expression is 3?

**Comment by ddobrev** [comment]
@IIFE it's impossible to know the Clang API by heart but I can guess. It most probably uses an expression (`clang::Expr*`) which can be evaluated to get the resulting value. For more details you'd need to check yourself. We have [some code to deal with expressions](https://github.com/mono/CppSharp/blob/master/src/CppParser/Parser.cpp#L3574) and [evaluation in particular](https://github.com/mono/CppSharp/blob/master/src/CppParser/Parser.cpp#L3658).

**Comment by tritao** [comment]
@IIFE This is not easily possible at the Clang level, at least not generally.

We've solved this issue by doing macro expression evaluation ourselves in the managed side.

See: https://github.com/mono/CppSharp/blob/master/src/Generator/Utils/ExpressionEvaluator.cs
https://github.com/mono/CppSharp/blob/7acd5809148ee7bdf843328d8e96f0b07b5bdf96/src/Generator/Library.cs#L113

It's possible the logic might need some tuning to handle your specific case.

**Comment by IIFE** [comment]
@tritao Thanks. Yes the `GenerateEnumFromMacros` call will eventually go through that evaluator, however it is in that evaluator where it's blowing up because it's not handling XORed macros. As you said the logic may need updating to handle that case.

**Comment by tritao** [comment]
By the way, I think `D3` is a logical OR (and not a XOR, which would be `ˆ` operator).

It should be simple to step through the expression evaluator and see where its blowing up.

**Comment by IIFE** [comment]
Yes sorry, meant to say logical OR. I'll try to write a minimal test and see the exact place where it blows up.
